storage/src/system/auction.rs

```diff
@@ -699,4 +699,70 @@ pub trait Auction:
             Err(Error::ValidatorNotFound)
         }
     }
+
+    /// Updates a `ValidatorBid` and all related delegator bids to use a new validator public key.
+    ///
+    /// This in effect "transfers" a validator along with its stake and all delegators
+    /// from one public key to another.
+    /// This method can only be called by the existing validator.
+    ///
+    /// The arguments are the existing validator's key and the new validator's key.
+    fn transfer_validator(
+        &mut self,
+        validator_public_key: PublicKey,
+        new_validator_public_key: PublicKey,
+    ) -> Result<(), Error> {
+        let validator_account_hash =
+            AccountHash::from_public_key(&validator_public_key, |x| self.blake2b(x));
+
+        // check that the caller is the existing validator
+        if !self.is_allowed_session_caller(&validator_account_hash) {
+            return Err(Error::InvalidContext);
+        }
+
+        // verify that a bid for existing validator exists
+        let validator_bid_addr = BidAddr::from(validator_public_key.clone());
+        let mut validator_bid = read_validator_bid(self, &validator_bid_addr.into())?;
+
+        // verify that a bid for new validator does not exist yet
+        let new_validator_bid_addr = BidAddr::from(new_validator_public_key.clone());
+        if self.read_bid(&new_validator_bid_addr.into())?.is_some() {
+            return Err(Error::TransferValidatorBid);
+        }
+
+        debug!("transferring validator bid from {validator_bid_addr} to {new_validator_bid_addr}");
+
+        validator_bid.with_validator_public_key(new_validator_public_key.clone());
+        self.write_bid(
+            new_validator_bid_addr.into(),
+            BidKind::Validator(validator_bid),
+        )?;
+
+        debug!("pruning validator bid {}", validator_bid_addr);
+        self.prune_bid(validator_bid_addr);
```

Comment:
I think we need to check two things for this bid before pruning it:
1. Are there unbonding records for this bid (to be paid out in future eras)
2. Does this bid correlate to a validator in the current era?
3. Does this bid correlate to a validator in the next era?

If the answer to all 3 is "NO" then we can prune it. However, if its "YES" to any or all of them we cannot prune at this time as that would cause the following issues in a later era:
1. When an unbond gets processed the logic wont be able to find the bid record it correlates to.
2. When the rewards get paid out for the current era or the next era, the auto-stake behavior wont be able to find the bid record. 

The unbond scenario is easy to address; you could roll thru any existing and write new records / prune old in the same fashion you are doing here, essentially replacing them with the updated information so that the future processing to pay them out will look for and find the new bid record.

The current / next era scenarios are a little trickier. What i recommend is to add a new variant to the BidKind enum to act as a bridge between the new and old records, written under the key of the original public key, and amend the look up logic in reward payout to interpret the new variant to look up the bid under the new public key:
pub enum BidKind {
    /// A unified record indexed on validator data, with an embedded collection of all delegator
    /// bids assigned to that validator. The Unified variant is for legacy retrograde support, new
    /// instances will not be created going forward.
    Unified(Box<Bid>),
    /// A bid record containing only validator data.
    Validator(Box<ValidatorBid>),
    /// A bid record containing only delegator data.
    Delegator(Box<Delegator>),
    **/// A bridge between an old bid and a new bid
    Bridge {
        // The previous public key for the bid.
        old_key: PublicKey,
        // The new public key for the bid.
        new_key: PublicKey,
        // the era this record was written
        era_id: EraId,
    },**
}

That bridge record can remain in gs allowing future forward or backward traversal to also follow the move as well.

Reply:
As an aside, we might decide to prune away such bridge records in some future era if it is no longer relevant (for instance, today the long pole is the unbonding delay which is currently 7 eras; we could prune the record away after the 7th era after it was written has elapsed as no future process would need it). Or we could run a garbage collection for no longer relevant bridge records as part of a protocol upgrade. However, I think it is also a valid option to allow such bridge records to remain in gs at tip allowing future forward or backward traversal to also follow the move as well. 

Either way, I don't think you need to concern yourself with this further consideration as part of this PR. I mention it here merely for context.

Reply:
in the case of redelegate, it attempts to take the stake and delegate it to a bid associated with a specified public_key ... if the bid holder associated with that public key has used this new function to change their public_key in the meantime the redelegation would fail. it should instead look for the bridge record and if found redelegate to the bid record associated to the new public key.

there's also an interaction in the actual unbond flow where the unbond will go to the account associated with the original public key which is ok but might be surprising to a bid holder who as shifted their bid to a new public key and thus a different account. that functionality (at the top of the unbond method) could be left as is as it is technically correct but might be surprising behavior for someone who has moved their staking to a new account associated to the new public key they used this new functionality to shift their bid to. Alternately, the unbond logic could be amended to look for a bridge record and if present use the new public key to look up the entity to get the main purse to transfer the unbonded stake into.

I can jump on a call w/ you next week and go thru it w/ you if desired.

Reply:
I think i see what you mean regarding the unbonding flow - we have to handle both a situation when the target validator has changed their bid and also a case of operators who unstaked partially and then changed their bid's public key during the unbonding delay.

The last aspect I'm uncertain about is more general - I tested an implementation of the new `BidKind` variant and I'm worried that using this approach might make the internal API inconvenient to use. What is mean is that there are  some methods defined for the `BidKind` enum and since this new variant is semantically quite different from existing ones (doesn't immediately represent an actual bid with a stake), we'd probably have to change most of them to return an `Option`. This would require additional boilerplate whenever using those methods and could sometimes be confusing (for example what does `inactive()` mean for a bridge record).
I also played around with an alternative approach of just adding the new key and era_id as optional fields on `ValidatorBid` itself. While simpler it'd probably also have some unintended consequences.

I agree it'd be great to meet up and discuss a preferred solution. We'll set up a call to make sure we are on the same page regarding this and the other PR.